src/reflection.ts

```diff
@@ -867,17 +876,21 @@ export class Parser {
         throw new Error(`Malformed message: could not find union type: '${discriminatorValue}'`);
       }
 
-      const subTable = rawLambda(table);
+      const subTable = deserializer[0](table);
       if (!subTable) {
         throw new Error(`Malformed message: missing union field table: '${field.name()}'`);
       }
-      return deserializer(subTable);
+      return deserializer[1](subTable);
     };
   }
 
   // eslint-disable-next-line @foxglove/prefer-hash-private
-  private createUnionDeserializers(fieldType: reflection.Type, readDefaults: boolean) {
-    const unionDeserializers = new Map<number, (t: Table) => Record<string, any>>();
+  private createUnionDeserializers(
```

Comment:
This now returns a tuple. The first value is a function to "read a sub-message" from the table. Previously this was the thing that was using `readTableLambda` and was not using the correct type index.